### PR TITLE
[Merged by Bors] - feat(Algebra): `Submodule R A` is algebra over `Ideal A`

### PR DESCRIPTION
**Pull request by alreadydone**
Also upgrade `Submodule.mapHom` and `Ideal.mapHom` to RingHoms and show the former is also an `AlgHom` (`Submodule.mapAlgHom`).

---

- [x] depends on: #17708 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by jcommelin** [comment]
I'm confused by the dependency, because it has a much larger diff then this PR itself...

File: Mathlib/Algebra/Algebra/Operations.lean [hunk 1]
```diff
@@ -163,6 +163,9 @@ theorem mul_le_mul_left (h : M ≤ N) : M * P ≤ N * P :=
 theorem mul_le_mul_right (h : N ≤ P) : M * N ≤ M * P :=
   AddSubmonoid.mul_le_mul_right h
 
+theorem mul_comm_of_commute (h : ∀ m ∈ M, ∀ n ∈ N, Commute m n) : M * N = N * M :=
+  toAddSubmonoid_injective <| AddSubmonoid.mul_comm_of_commute h
```
Comment (alreadydone) [inline Mathlib/Algebra/Algebra/Operations.lean:167]: This won't work without the dependency PR which makes `(M * N).toAddSubmonoid` defeq to `M.toAddSubmonoid * N.toAddSubmonoid`.

**Comment by alreadydone** [comment]
!bench

File: Mathlib/RingTheory/Ideal/Operations.lean [hunk 2]
```diff
@@ -1260,6 +1259,39 @@ theorem set_smul_top_eq_span (s : Set R) :
...
+  map_mul' := (Submodule.map_mul · · <| Algebra.ofId R A)
+  map_zero' := map_bot _
+  map_add' := (map_sup · · _)
+  commutes' I M := mul_comm_of_commute <| by rintro _ ⟨r, _, rfl⟩ a _; apply Algebra.commutes
+  smul_def' I M := le_antisymm (smul_le.mpr fun r hr a ha ↦ by
+    rw [Algebra.smul_def]; exact Submodule.mul_mem_mul ⟨r, hr, rfl⟩ ha) (Submodule.mul_le.mpr <| by
+    rintro _ ⟨r, hr, rfl⟩ a ha; rw [Algebra.linearMap_apply, ← Algebra.smul_def]
+    exact Submodule.smul_mem_smul hr ha)
+
+/-- `Submonoid.map` as a `AlgHom`, when applied to an `AlgHom`. -/
```
Comment (jcommelin) [inline Mathlib/RingTheory/Ideal/Operations.lean:1280]:
```suggestion
/-- `Submonoid.map` as an `AlgHom`, when applied to an `AlgHom`. -/
```

File: Mathlib/RingTheory/Ideal/Operations.lean [hunk 3]
```diff
@@ -1260,6 +1259,39 @@ theorem set_smul_top_eq_span (s : Set R) :
...
+@[simps!] def mapAlgHom (f : A →ₐ[R] B) : Submodule R A →ₐ[Ideal R] Submodule R B where
+  __ := mapHom f
+  commutes' I := (map_comp _ _ I).symm.trans (congr_arg (map · I) <| LinearMap.ext f.commutes)
+
+/-- `Submonoid.map` as a `AlgEquiv`, when applied to an `AlgEquiv`. -/
```
Comment (jcommelin) [inline Mathlib/RingTheory/Ideal/Operations.lean:1285]:
```suggestion
/-- `Submonoid.map` as an `AlgEquiv`, when applied to an `AlgEquiv`. -/
```

**Review by jcommelin — COMMENTED** [review]
bors d+

**Comment by alreadydone** [comment]
Thanks for the reviews! Sorry I got the commit message backwards 🤣

**Pull request closed without merging**